bemani/backend/jubeat/clan.py

```diff
@@ -1113,21 +1114,26 @@ def format_scores(self, userid: UserID, profile: ValidatedDict, scores: List[Sco
             ex_cnt = data.get_int_array('ex_cnt', 3)
             points = data.get_int_array('points', 3)
 
+            # This means that we already assigned a value and it was greater than current
```

Comment:
Not sure I agree with the philosophy on this. Did clan separate the scores or was "hard mode" just a tracking flag? I know festo has them separate. Trying to figure out what to do about existing support. I think the change for adding hard mode is going to be where we spend a lot of focus, to ensure its right.

Reply:
The  game sends the hard mode flag as far back as saucer. From what I could tell, the "correct" behavior is to send the highest of the two scores rather than filter them outright

Reply:
What about highest of other things? The score save method for every game does a merge of the best properties (combo indicator, ghost, score, rate, etc. So if you get a full combo but worse score, your record still shows your best score AND full combo) like the real network does, so your stored score is always the best of each individual thing. If we are going to return the best of either normal or hard mode scores, I think we need to do this merging here as well to preserve the functionality.

Reply:
To be completely honest with you, I forgot about that edge case and now I realise I have no idea how to deal with that.

Reply:
Yeah, providing support for score portability backwards and forwards for games that never intended that sort of behavior all while preserving the expected behavior for each game is **HARD**. I've spent many sleepless nights getting that stuff just right for games. Believe me, its a royal pain whenever they throw a curveball like this. If you look at the data providers for fetching scores from remote servers as well as local mysql (when you have specified a remote bemapi-compatible server as a data source), there's code in there that does this merging so that if, for instance, you get a full-combo on your friend's network but have a higher score on yours, since they're linked it will still show correctly on your game. We might want to generalize that merging code into a library, call it from the API fetch methods, and then in jubeat older than festo fetch the normal and hard mode scores and do that merge before presenting to the client.

Generalizing those merges would be beneficial for another reason: people sometimes play on two cards, and want to merge their accounts. Merging two accounts is really the act of pointing all cards in account A to account B, dropping account A from the user tables, copying over all the history entries and then merging all of the high scores. Everything there is trivial, except merging. Hence, why there's no admin function for doing that merge (yet). If we had that generalized, it would be one step closer to implementing a true "merge two accounts".